[Console] Bash completion integration

wouterj opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #38275
| License       | MIT
| Doc PR        | -

This is a first shot at implementing interactive bash completion support in the Console component. Besides completing option and command names, commands can override `Command::complete()` to implement completing values.

I've added an example code to the `secrets:remove` command, which now supports autocompletion quite nicely: ![render1630315116886](https://user-images.githubusercontent.com/749025/136708284-bf2e4c12-7cb7-4d5e-9c8d-68bcdca6fd7c.gif)

And support for other applications using `symfony/console` is automatically included (if the autocompletion script is installed for that specific application): 
![render1630314403752](https://user-images.githubusercontent.com/749025/136708323-dfbccb77-dcbd-4d1e-8bb5-85b88f0b358b.gif)

This PR only implements Bash completion. Zsh and Fish have much more sophisticated completion systems, but I propose to keep those for a future PR if there is a need for this.

### How it works

1. A bash completion function (`_console`) is defined by `bin/console completion bash` for the `console` command (name of the "binary" file)
2. This completion function calls the local `bin/console _complete` command to come up with suggestions
3. Bash parses these suggestions and shows them to the user.

This has one drawback: the `_console` function is defined globally only once. This means we cannot easily change it, as it would break if you run different Symfony versions. We should probably add versioning (e.g. `bin/console _complete --version=1.0`) and don't suggest anything if the version doesn't match.

<s> **Maybe it would be safer to mark this feature as experimental and target 6.0, to allow us to fine tune the shell related sides over the lifespan of 6.x?** </s> https://github.com/symfony/symfony/pull/42251#issuecomment-890925658

### Steps to test yourself

Load this PR in your project, open a bash shell and run this command to "install" completion for this project:

```
bin/console completion bash > /etc/bash_completion.d/console
````

Then reload the bash shell and enjoy autocompletion.

### TODO

* [x] Autocompleting in the middle of the input doesn't work yet (i.e. `bin/console --en<TAB> cache:clear`)
* [x] Better error handling
* [x] Add a `bin/console completion` command to dump the `_console` file, so users can install this locally
* [x] Add some versioning, to allow us to change the `_console` file in the future
* [x] <s>See how we can better support standalone usage (e.g. Composer)</s> Tested on Laravel's artisan, works flawlessly

wouterj committed:
[Console] Bash completion integration

wouterj committed:
[FrameworkBundle] Add CLI completion to secrets:remove

Nyholm commented:
Thank you. 

This looks super cool. =)

wouterj commented:
Thanks :)

I've updated this PR to complete most of the important TODO items (and also updated the usage in the PR description). It still needs more unit tests (I used a local shell script to test things, I need to convert it to unit tests) and I need to add a line to the CHANGELOG. Other than that, I think this in a final state, so ready for reviews :)

wouterj commented on src/Symfony/Component/Console/Resources/completion.bash:
```diff
@@ -0,0 +1,31 @@
+    for w in ${words[@]}; do
+        completecmd+=(-i "'$w'")
+    done
+    #echo
+    #echo ${completecmd[@]}
```
```suggestion
```

(left over "debug" statement)

wouterj commented on src/Symfony/Component/Console/Completion/Output/BashCompletionOutput.php:
```diff
@@ -0,0 +1,35 @@
+        $options = [];
+        foreach ($completion->getOptionSuggestions() as $option) {
+            $options[] = '--'.$option->getName();
+            /*if ($shortcut = $option->getShortcut()) {
+                $options[] = '-'.$shortcut;
+            }*/
```
```suggestion
```

Can probably be removed, I think it doesn't make sense to complete short option names.

OskarStark commented:
Should we target `6.0` and mark this big new feature as `@experimental`?

Nyholm commented:
My general opinion is that we mark too many classes experimental. 

But I let @wouterj decide in this specific case.

wouterj commented:
The public API side of this feature is quite minimal (one method in `Console` which needs full BC and 2 classes that we should mark as `@final`). The feature itself is purely DX (i.e. it's not that big of a deal if it stops working), so I agree with not marking this as experimental.

Adding support for ZSH might break things, but a community member already stepped up to provide a PR for this if this is merged, so this can be caught before the 5.4 release. Other than that, all complexity of this feature is in the `CompletionInput::bind()` method, which will all be normal "bug fixes".

---

I guess the lack of reviewers in this PR indicates that it's probably a bit difficult to review. Some hints:

- The main logic is in `CompleteCommand` and `CompletionInput`. The command receives an array of input "words" (similar to `$argv`) and the index in that array of the current cursor position (to know where completion is requested). Based on this, `CompletionInput` has to provide the correct type, value, option name, etc. That is the core logic of this PR (see also its tests).
- From a DX side, see `SecretsRemoveCommand::complete()`, `Command::complete()` and `Application::complete()` for PHP usage of this feature.
- If you're into bash, consider reviewing `completion.bash` and `BashCompletionOutput`. That latter is not doing much (only providing completion options as space-separated values). The first one is a bit more complex than the basic bash completion function, in order to simplify working with colons in the command names and `=` signs in between option name and value.

dkarlovi commented:
@wouterj I'm looking for a simple way to test this in my actual app, IMO it would be most beneficial to figure it all out. I'll provide my findings if I find a way to do that.

Other than that, a cool sample might be to make a fork of Composer with this enabled. It might make it simple to visualize once you can test it for yourself easily.

dkarlovi commented:
Also, this is an absurdly valuable feature, thank you a lot for working on this! 😍

wouterj commented:
@dkarlovi if you have an app running Symfony 5.3, you can try to clone my fork locally and use the `link` utility:

```
git clone https://github.com/wouterj/symfony -b issue-38275/shell-autocomplete
cd symfony
./link /path/to/project
```

And then use the `eval $(bin/console completion bash)` in your app.

dkarlovi commented:
@wouterj playing with this currently in my own app. It seems there's some fragility, for example:

## Arguments not properly marked as mandatory

```
  Symfony\Component\Console\Completion\CompletionInput::fromTokens(): Argumen  
  t #2 ($currentIndex) must be of type int, null given, called in /home/dalib  
  or.karlovic/Development/RandD/symfony-autocomplete/src/Symfony/Component/Co  
  nsole/Command/CompleteCommand.php on line 109    
  ```
  
## Argument types not properly validated / cast

  ```
    Symfony\Component\Console\Completion\CompletionInput::fromTokens(): Argumen  
  t #2 ($currentIndex) must be of type int, string given, called in /home/dal  
  ibor.karlovic/Development/RandD/symfony-autocomplete/src/Symfony/Component/  
  Console/Command/CompleteCommand.php on line 109
  ```
  
  ## Arguments boundaries not properly validated

  ```
  In CompletionInput.php line 198:
                                  
  [ErrorException]                
  Warning: Undefined array key 9
  ```

dkarlovi commented:
@wouterj which Bash did you use to test this? For me, if I test with `bin/console`, it doesn't work, but `./console` works, but need to change CWD obviously.

To get it to work, I needed to change it like this:

```
complete -F _sf_console bin/console
```

Also, the `eval` trick you gave didn't register the completion for me (checked with `complete -p console`), had to pipe it to a file and source it

```
bin/console completion bash > foo.sh
source foo.sh
```

dkarlovi commented on src/Symfony/Bundle/FrameworkBundle/Command/SecretsRemoveCommand.php:
```diff
@@ -80,4 +82,19 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+            CompletionInput::TYPE_ARGUMENT_VALUE === $input->getCompletionType()
+            && 'name' === $input->getCompletionName()
+        ) {
+            $vault = $input->getOption('local') ? $this->localVault : $this->vault;
```
`$vault` unused here.

dkarlovi commented on src/Symfony/Component/Console/Resources/completion.bash:
```diff
@@ -0,0 +1,29 @@
+    __ltrim_colon_completions "$cur"
+}
+
+complete -F _sf_{{ COMMAND_NAME }} {{ COMMAND_NAME }}
```
Needed 

```bash
complete -F _sf_console bin/console
```

to get it to work on Bash 5.1.0

dkarlovi commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,136 @@
+            $completionOutput = new $completionOutput();
+            $completionOutput->write($completion, $output);
+        } catch (\Throwable $e) {
+            if ($this->logger) {
```
If your logger is set to STDOUT, you get the error as the completion here.

wouterj replied:
Hmm, that's not great of course (I explicitly used this logger to avoid having the exception message as completion output).

What do you think about having a debug flag, like you proposed in https://github.com/symfony/symfony/pull/42251#issuecomment-893337446 and then always directly print the exception message to STDERR (at least, I expect this won't be picked up by compgen)?

dkarlovi commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,136 @@
+        $currentIndex = $input->getOption('current');
+        $completionInput = CompletionInput::fromTokens(array_map(
+            function (string $i): string { return trim($i, "'"); },
+            $input->getOption('input')
```
I've noted these need to be validated better in the comments.

wouterj replied:
Are you maybe able to show me how to produce the errors you've shown? I haven't received those errors while writing/testing this

dkarlovi replied:
Sure, try

```
bin/console _complete -v -s bash

bin/console _complete -v -s bash -c 1

bin/console _complete -v -s bash -c 2 -i a

bin/console _complete -v -s bash -c a -i 2
```

dkarlovi commented:
One thing which might be neat to have is to provide a flag for the dumper, something like

```
bin/console completion bash --verbose-completion
```

or similar, where it would create additional code in there to help you debug it if you're working on it. Might be too niche, maybe wait for use cases, but as I was testing this, I needed to edit the completion function to add them myself.

dkarlovi commented:
> See how we can better support standalone usage (e.g. Composer)

Typically on systems which embrace this, they ship the completion file with the tool, example

```
$ rpm -ql kubernetes-client | grep comple
/usr/share/bash-completion/completions/kubectl
```

This automagically enabled completion for end users. If Composer integrated this, it would be able to generate its own completion file (already here) and the people packaging it would build the completion file as they're building the package. Then it's up to Composer itself to autocomplete everything properly, which would need to be done per command basis (for example `composer require <tab>`, etc).

This seems like exactly the correct first step, Composer would need to embrace this feature once released.

wouterj commented:
> @wouterj which Bash did you use to test this? For me, if I test with bin/console, it doesn't work, but ./console works, but need to change CWD obviously.

Oh, I see. I used `bin/console completion bash > /etc/bash_completion.d/console` when testing this patch (on bash 5.1.8). I expected `eval()` to work but apparently it doesn't.
Also seems like sourcing the file separately (like you show) requires `./console <tab>`, while putting it in `bash_completion.d` works with `bin/console <tab>` :thinking:

wouterj commented:
Thanks for testing this thoroughly @dkarlovi !

I've added more validation to the command (still need to write test for this command) and updated the installation instructions. I also found some more bugs for the `CompletionInput` by adding some more tests. I'll need to dig a bit deeper in the setup (i.e. bamarni's package suggests `eval`, so I'm not sure why it doesn't work here).

chalasr commented:
Looks super promising!
Does this support zsh? I couldn’t manage to make it work on my machine yet, I probably need to try harder.

wouterj commented:
No, currently only bash is supported (ZSH has its own completion system and some basic bash completion support, but not enough to support this file - due to the magic required to make colons work).

I don't want to add more implementations in this PR, it's big enough on its own imho. A community member already volunteered to write a zsh script after this is merged (and I also have some basic script ready, as I started working on this PR in zsh before switching to bash). So I'm certain that we will have bash and zsh support before the release (and, if someone has time, fish support, but I have no personal need for that).

At this moment, I'm using the `bash` command in zsh to open a bash shell and test this :)

wouterj commented:
Updated this PR to fix the 2 open comments:

- It turns out `eval $(...)` doesn't work when run in an open shell. You'll have to add it to your `.bashrc`/`.bash_profile` instead
- I've removed the logger and implemented a basic custom debug helper.
  1. You can enable debug mode by running `export SYMFONY_COMPLETION_DEBUG=1` once in your shell
  2. Then open another terminal window and run `bin/console completion --debug`.
  3. Enjoy live debugging information when trying to use the bash completion in the shell of (1):
     ![image](https://user-images.githubusercontent.com/749025/131250086-1178da81-16ea-410e-9d19-27d63eaad5c7.png)

wouterj commented:
I'm unsure how to fix the phpt test of runtime on low-deps: increase the minimum version of `symfony/console`?

Other than that, the build is :green_book:  (fabbot is a false positive)

ro0NL commented on src/Symfony/Component/Console/Completion/CompletionInterface.php:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+interface CompletionInterface
```
Is the contract needed?

wouterj replied:
Strictly, no (given that it's implemented by both main contracts of this component; Application and Command). But I thought it would be nice to make sure both application and command use the same completion contract.

ro0NL replied:
I'd keep it an implementation detail for now. For example, can we also think of a way `parent::complete` isnt needed?

Either way ... :+1:  :)

wouterj replied:
I've slightly updated the PR based on your suggestion:

- I've moved the input option completion to the `CompleteCommand` directly (removing the parent call)
- `Command` now longer implements `complete()` and commands need to implement this interface if they want to add custom completion

Neirda24 commented:
> @wouterj which Bash did you use to test this? For me, if I test with `bin/console`, it doesn't work, but `./console` works, but need to change CWD obviously.
> 
> To get it to work, I needed to change it like this:
> 
> ```
> complete -F _sf_console bin/console
> ```
> 
> Also, the `eval` trick you gave didn't register the completion for me (checked with `complete -p console`), had to pipe it to a file and source it
> 
> ```
> bin/console completion bash > foo.sh
> source foo.sh
> ```

I find this to be the best approach. It allows for multiple completion version on different symfony projects. Simple and flexible but need to be added to the default `.gitignore` of symfony.

wouterj commented on src/Symfony/Component/Runtime/composer.json:
```diff
@@ -21,7 +21,7 @@
     },
     "require-dev": {
         "composer/composer": "^1.0.2|^2.0",
-        "symfony/console": "^4.4|^5.0|^6.0",
```
@nicolas-grekas would you agree with this change? (it removes quite some versions from the test)

There is a `.phpt` test that must check different outputs for <5.4 and >=5.4. I don't know of any way to conditionally run `.phpt` tests based on class existence (in PHPunit, I would normally create 2 tests and use `markTestAsSkipped()`).

_for some reason, the tests now use `6.0.x` as lowest version. so it's still red, but it will be green once this PR is merged_

nicolas-grekas replied:
I'd rather relax the test case on 5.4

derrabus replied:
@wouterj You can skip PHPT tests using a `SKIPIF` script. If that script outputs anything, the main test won't be executed. Example:

https://github.com/symfony/symfony/blob/a98735b7e9885dcb3f270fa95a4565164322c316/src/Symfony/Component/ErrorHandler/Tests/phpt/fatal_with_nested_handlers_pre81.phpt#L3-L4

wouterj commented:
Made some final changes and added some more docs (phpdoc + command help about installation).

Fabbot errors are both false positives. Imho, this is now ready.

nicolas-grekas commented on src/Symfony/Component/Console/Tests/Fixtures/application_run1.txt:
```diff
@@ -12,5 +12,6 @@ Options:
 Available commands:
-  help  Display help for a command
-  list  List commands
+  completion  Dump the shell completion script
```
or shouldn't we hide the command (and document it elsewhere?), so that it doesn't pollute the list, since it's a use-once command?

Neirda24 replied:
Use once every time we change something. I think it should be kept in the listing. Or create a composer lock like behaviour. Detecting that the dump should be re-do when using `console`

nicolas-grekas replied:
If we somehow manage to make this autoupdateable, we could hide the command.

dunglas replied:
Having it listed also improves the discoverability of this feature.

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Command/SecretsRemoveCommand.php:
```diff
@@ -80,4 +83,15 @@ protected function execute(InputInterface $input, OutputInterface $output): int
+
+    public function complete(CompletionInput $input, Completion $completion): void
+    {
+        if ($input->mustSuggestArgumentValuesFor('name')) {
```
`if (!...) return;`

nicolas-grekas commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,191 @@
+        $this
+            ->setName(ltrim(self::$defaultName, '|'))
+            ->setDescription(self::$defaultDescription)
+            ->setHidden(true)
```
can't we remove the 3 above lines?

wouterj replied:
The command is registered by the `Application` class, not by the compiler pass. So I don't think the static logic works here.

nicolas-grekas replied:
it should afaik, there's logic for that in the base constructor

nicolas-grekas commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,191 @@
+
+    protected function initialize(InputInterface $input, OutputInterface $output)
+    {
+        $this->isDebug = (bool) getenv('SYMFONY_COMPLETION_DEBUG');
```
`filter_var(getenv('SYMFONY_COMPLETION_DEBUG'), \FILTER_VALIDATE_BOOLEAN)`

nicolas-grekas commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,191 @@
+            ->setHidden(true)
+            ->addOption('shell', 's', InputOption::VALUE_REQUIRED)
+            ->addOption('current', 'c', InputOption::VALUE_REQUIRED)
+            ->addOption('input', 'i', InputOption::VALUE_REQUIRED | InputOption::VALUE_IS_ARRAY)
```
maybe worth a few explanations?

nicolas-grekas commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,191 @@
+            }
+
+            $shell = $input->getOption('shell');
+            if (!$shell) {
```
to be merged with the previous line (same on similarly split lines) :)

wouterj replied:
Do we really want to enforce this CS rule? Imho, it massively reduces code readability for little gain

nicolas-grekas replied:
to my eyes it improves readability :)

nicolas-grekas commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,191 @@
+                        '  Completing using the <comment>'.\get_class($command).'</> class.',
+                        '  Completing <comment>'.$completionInput->getCompletionType().'</> for <comment>'.$completionInput->getCompletionName().'</>',
+                    ]);
+                    if ($compval = $completionInput->getCompletionValue()) {
```
what if compval is `0`?

nicolas-grekas commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,111 @@
+
+Dump the script to a global completion file and restart your shell:
+
+    <info>%command.full_name% bash | sudo tee /etc/bash_completion.d/${commandName}</info>
```
What about the `SHELL` env var instead of forcing to add `bash` explicitly?

I see that this is the case already, so: let's remove bash from the examples to make them more generic?

nicolas-grekas commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,111 @@
+
+    <info>%command.full_name% bash | sudo tee /etc/bash_completion.d/${commandName}</info>
+
+Or dump the script to a local file and source it:
```
Would it make sense to always generate this script as part of a cache warmer?
Could that ease with installing/updating the completion script?

dkarlovi replied:
> Would it make sense to always generate this script as part of a cache warmer?

No, because installing the completion script is system-dependent, **the same completion script is used for multiple Symfony apps** on your system. Unless the completion is bundled with the app like I've shown in the `kubectl` example, the user needs to enable it for themselves in some way. The dynamic of updating this script will not be the same as the dynamic of building the DIC.

Also, note that, if the script is indeed installed in this manner (system-wide), cache warmer will not have permissions to update it. On some systems, a user-localized completion is supported by default, here's the snippet from Fedora:

```bash
if [ "x${BASH_VERSION-}" != x -a "x${PS1-}" != x -a "x${BASH_COMPLETION_VERSINFO-}" = x ]; then

    # Check for recent enough version of bash.
    if [ "${BASH_VERSINFO[0]}" -gt 4 ] ||
        [ "${BASH_VERSINFO[0]}" -eq 4 -a "${BASH_VERSINFO[1]}" -ge 2 ]; then
        [ -r "${XDG_CONFIG_HOME:-$HOME/.config}/bash_completion" ] &&
            . "${XDG_CONFIG_HOME:-$HOME/.config}/bash_completion"
        if shopt -q progcomp && [ -r /usr/share/bash-completion/bash_completion ]; then
            # Source completion code.
            . /usr/share/bash-completion/bash_completion
        fi
    fi

fi
```

Neirda24 replied:
What about generating the file during cache warming and put in a dedicated folder of the app (with gitignore) and propose a script on the side (maybe through symfony cli) that could watch current working directory changes and auto source / unsource the file ? So it stays app specific and a good DX.

nicolas-grekas commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,111 @@
+<comment>Dynamic installation
+--------------------</comment>
+
+Add this add the end of your shell configuration file (e.g. <info>"~/.bashrc"</info>):
```
this means that the script is going to be generated every time a completion is attempted?
won't that kill the DX?

wouterj replied:
It means the script is generated every time you open a new shell.

The disadvantage is a slight loss of performance when opening the new shell. The advantage is always having the up to date completion script.

dkarlovi replied:
But which copy of console do you point it to?

Speaking of, the completion script should catch the exception where the completion command is unrecognized. Namely, if you're trying to complete on an old Symfony project.

Neirda24 replied:
It shouldn't be put into the `.bashrc` What about multiple symfony projects ? Couldn't we have like a dump inside the project itself and then a bash script in the bashrc that check for the existence of this file (maxdepth=2 for example) or even an env variable that contains the target to source ?

One last thing : I'm not completely sure how it will work yet but it will require php to autocomplete or just to dump the completion script ? If it requires php to autocomplete won't that kill DX when using docker for example ?

nicolas-grekas commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,111 @@
+        if ($input->getOption('debug')) {
+            $this->tailDebugLog($commandName, $output);
+
+            return 0;
```
self::SUCCESS for consistency

nicolas-grekas commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,111 @@
+        $completionFile = __DIR__.'/../Resources/completion.'.$shell;
+        if (!file_exists($completionFile)) {
+            ($output instanceof ConsoleOutputInterface ? $output->getErrorOutput() : $output)
+                ->writeln(sprintf('<error>Detected shell "%s", which is not supported by Symfony shell completion.</>', $shell));
```
let's tell what is supported?

nicolas-grekas commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,111 @@
+        return self::SUCCESS;
+    }
+
+    private function guessShell(): string
```
not sure the method is worth it :)

wouterj replied:
Yeah, not at the moment. I wanted to include more logic. E.g. the `$SHELL` var shows your main shell variant, not the current variant. So we can be more fancy here and check currently existing vars to see what shell is used at the moment.

nicolas-grekas commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,111 @@
+
+    private function guessShell(): string
+    {
+        return basename($_SERVER['SHELL']);
```
`basename($_SERVER['SHELL'] ?? '');`?

nicolas-grekas commented on src/Symfony/Component/Console/Completion/Completion.php:
```diff
@@ -0,0 +1,87 @@
+    /**
+     * Add a suggested value for an input option or argument.
+     */
+    public function suggestValue(string $value): self
```
let's add `@return $this` everywhere it makes sense

nicolas-grekas commented on src/Symfony/Component/Console/Input/ArgvInput.php:
```diff
@@ -68,17 +68,22 @@ protected function parse()
+        }
+    }
+
+    protected function parseToken(string $token, bool &$parseOptions): void
```
can we return something instead of passing by ref?

nicolas-grekas commented on src/Symfony/Component/Console/Tests/Descriptor/ObjectsProvider.php:
```diff
@@ -74,6 +74,10 @@ public static function getCommands()
     {
+        // The help text depends on the dynamic "PHP_SELF" value, which cannot
+        // be tested against. This makes the value predictable.
+        $_SERVER['PHP_SELF'] = 'app/console';
```
Shouldn't we restore this, eg in tearDownAfterClass?

wouterj replied:
Solved it in a nicer way :) (as there were now even more dynamic things)

stof replied:
shouldn't you removed this then ?

nicolas-grekas reviewed (commented):
Nice work!
Here are some random comments :)
I'm mainly looking to smoothen the DX around installing+updating this.
Eg right now we tell about 3 ways to install this. Having only one would be better if possible. I made some suggestions that might help in this direction, hopefully.

dkarlovi commented on src/Symfony/Component/Console/Resources/completion.bash:
```diff
@@ -0,0 +1,33 @@
+
+    local sfcomplete=$(${completecmd[@]})
+    if [ $? -ne 0 ]; then
+        echo $sfcomplete
```
Noted this before: if you use this feature on a Symfony app where complete is not available (older Symfony), it should just not work. The error shouldn't be presented as completion options.

wouterj replied:
I would love to, but I don't know how. There is no special exit code and the errors are written to stderr.

If you have a solution, please let me know :)

dkarlovi replied:
What happens if you try to complete on a copy of Symfony which doesn't have completion? Say, <= 5.3? My comment is just this should fail somewhat gracefully, it may already work.

wouterj replied:
Yeah, it's not graceful at the moment :cry: 

![image](https://user-images.githubusercontent.com/749025/132834967-cea2ee64-035a-49e4-80dd-fa4e13262b39.png)

dkarlovi replied:
Can we maybe capture the STDERR and check it?

dkarlovi commented:
@wouterj just saw [the screenshot example](https://github.com/symfony/symfony/pull/42251#issuecomment-907782832) of the debugging and I must say, that's straight up **amazing** work!

fabpot commented:
https://github.com/posener/complete/tree/master is an interesting Go project that does the same for Bash, fish, and zsh.

wouterj commented:
To make my intentions with this PR clear: the library mentioned by @fabpot takes an interesting approach of "predicting" values, or something or nothing. I'm not sure yet whether the approach in this PR, or the approach by the bundle would be optimal.

There is also an open discussion still how to better support multiple symfony projects on the same machine: https://github.com/symfony/symfony/pull/42251#discussion_r703313508

All in all, I think the API of this feature is too undefined to merge it as experimental in 6.0. And given my open source attention is now on the security documentation, I won't have any time to polish it before 5.4/6.0. So I think we have to postpone this to 6.1 unfortunately

dkarlovi commented:
@wouterj let me preface this by again saying thanks a lot for the effort of working on this!

From my POV, this code actually looks pretty much feature complete, the new API footprint is IMO small and seems good to me, not something obviously wrong with it.

> There is also an open discussion still how to better support multiple symfony projects on the same machine

From my POV this is very clear, especially since there's loads of real examples how this ends up working in practice. To possibly further the discussion, let me outline my understanding here.

# Multiple projects

- the completion script is **global** to the shell, user, system, you can only have one completion script running at a time, for simplicity sake here let's say it's system-wide, it's NOT related to your project
- the completion command is **local** to the Symfony app you're working with, let's say we have three apps called `sf53`, `sf54` and `sf61`
- the same completion script (again, global) would need to work for all three (again, local) completion commands, as the user is moving about their system, the completion script can exist even if there are **zero** Symfony projects on the system, for example, Symfony might have been installed via [Symfony RPM packages](https://centos.pkgs.org/8/remi-x86_64/php-symfony-common-2.8.52-2.el8.remi.noarch.rpm.html) which can contain the completion script generated with that version of Symfony
- this means that **the same completion script** installed in `/usr/share/bash-completion/completions/console` will trigger completion for `~/sf53/bin/console`, `~/sf54/bin/console`, ``~/sf61/bin/console``
- this means the completion script generation is shipped **as a convenience** in each project, it's not meant to be run each time you switch projects, it's meant to be run once (or zero times) and then work with all of your existing projects (where possible, of course), it is absolutely not meant to be built and installed from each project or with each container build, it shouldn't be hidden or removed, other projects ship theirs without qualm:
```shell
$ rpm -ql kubernetes-client | grep completion
/usr/share/bash-completion/completions/kubectl
/usr/share/man/man1/kubectl-completion.1.gz
$ kubectl | grep completion
  completion    Output shell completion code for the specified shell (bash or zsh)
```

The completion script is more similar in nature to the `symfony` CLI than `bin/console`. 

# Compatibility

What could be a good idea is to generate the completion script in a way where it's saying the completion command which version generated it. Something like

```
bin/console _complete -v -s bash -c a -i 2 --symfony=5.4.2
```

This gives the completion command an opportunity to either patch the behavior across versions (which is unlikely to even be required since it's so simple) or signal to the completion script to hint to the user it needs to be refreshed.

For example:
1. `bin/console _complete -v -s bash -c a -i 2 --symfony=6.1.1`
2. there's a known bug in the completion script generated with this version
3. completion command sees the version it was invoked with, echoes the message and returns `255` or similar exit code
4. completion script is generated in a way to know what to do with this when it happens from the start

# Remaining steps

IMO, we should:
- [x] add this version param to the completion command, it doesn't need to do anything yet
- [x] tweak completion script generator to add the version into the script
- [x] tweak completion script generator to handle the "signal" `255` described above
- [x] tweak completion script generator to handle errors from Symfony 5.3 gracefully

Other than that, this seems pretty complete to me.

wouterj commented:
Thank you for writing it down so clearly @dkarlovi, much appreciated.

I've just pushed the changes you listed. So this is a mergeable feature now. I do fully agree with @dkarlovi that having a global shell script and a runtime command to provide the suggestions is the best solution here. There must always be one manual action to hook into the local shell completion, so there must always be some shell script installation process.

Making changes to use cache warmers to generate the suggestions will make the suggestions potentially outdated and it will not work with all standalone Console applications (Composer, Behat, artisan, etc). Note that performance is not really a big thing here, I'm on a 8 year old laptop with an HDD and the suggestions come pretty quick for me (not slower than other completions).

I'll leave this to the core team to review and decide if it's worth for 5.4, 6.0 or 6.1 :) The build failures are intentional: fabbot has false positives, and Psalm discovers an unlikely edge case that is not worth fixing imho

stof commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,195 @@
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class CompleteCommand extends Command
```
should be final IMO

stof commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,117 @@
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class DumpCompletionCommand extends Command
```
should be final

dkarlovi commented:
> Making changes to use cache warmers to generate the suggestions will make the suggestions potentially outdated and it will not work with all standalone Console applications

IMO this could be the next step after this feature is merged: allowing apps to "bake in" some static completions which could then be resolved ASAP. If you're resolving command/parameter names (especially in prod mode where they're not expected to change), there might be some opportunity to resolve them without invoking the complete command lifecycle.

 But, I'd do that once we see the issue, can profile it and see where it's coming from, if at all. I wouldn't solve the feature starting with the optimization.

dkarlovi commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,195 @@
+
+            //    $output->writeln($message.' Install the Symfony completion script again by using the "completion" command.');
+
+            //    return 126;
```
Constant?

wouterj replied:
IIRC, Fabien isn't a big supporter for these type of constants (like SUCCESS and FAILURE). So as this has a single use, let's not introduce a constant for it (no real big benefit, so not worth a discussion :wink: ).

dkarlovi commented on src/Symfony/Component/Console/Resources/completion.bash:
```diff
@@ -0,0 +1,36 @@
+        COMPREPLY=($(compgen -W "$sfcomplete" -- "$cur"))
+        __ltrim_colon_completions "$cur"
+    else
+        if [[ "$sfcomplete" != *"Command \"_complete\" is not defined."* ]]; then
```
Would you not test for code `126` here? This is why I said it could be a constant: you could inject it here and have special casing for it.

wouterj replied:
Unfortunately, we can't. Applications using <5.4 aren't aware of having to use a special exit code for the `_complete` command. The only thing we can do to fail gracefully for these apps is to check the stderr body.

dkarlovi replied:
Sorry, I marked the wrong line.

I mean, SF 5.4+ (assuming this gets merged) apps return `126` which means specifically "notify the user", so I'm wondering what should the completion script do with the 126 code to notify the user, should it have some special casing?

I don't mean handling the general error for <5.4 apps, that part is clear.

dkarlovi replied:
To clarify further: when I generate my completion script with 5.4 and forget about it, what does it do in the future when Symfony 6.2 returns `126`?

wouterj replied:
I tested this while writing the future, you'll get the stdout of the `_complete` command printed to your console (using stderr):

```
$ bin/console <TAB>
Completion script version is not supported ("5.4.0" given, ">=6.2" required). Install
the Symfony completion script again by using the "completion" command.
```

The error message is not good yet, but I think this is the best we can get.

This works the same for any other failure of the `_complete` command (except from "command not defined"), which I think gives us the best flexibility to adding more helpful errors in the future in the PHP command without having to update the bash script.

dkarlovi reviewed (approved):
LGTM.

dkarlovi commented:
One more thing which came up in discussion in Slack: assuming you want to enable Linux distro packagers to build the completion at package's build time ([example for `kubectl`](https://src.fedoraproject.org/rpms/kubernetes/blob/rawhide/f/kubernetes.spec#_211)), what would be the suggested way to do that for, let's say, Composer (assuming they get a chance to integrate this work)?

Would Composer (or any other standalone tool) simply integrate the `complete` command to generate their own completion shell script and that's it?

Here's Composer's current Fedora SPEC file (used to build the RPMs):
https://src.fedoraproject.org/rpms/composer/blob/rawhide/f/composer.spec

Edit: maybe @remicollet can weigh in here, hopefully he'll not mind the mention. :crossed_fingers:

wouterj commented:
> Would Composer (or any other standalone tool) simply integrate the complete command to generate their own completion shell script and that's it?

Aside from whether this is the wanted behavior: this is currently possible (the Laravel demo in the PR description was made just by adding a `complete()` method to one Laravel command).

I guess from a package maintainer perspective, we would need to support a partially static completion script for this to be considered (e.g. generate a script based on the `InputDefinition` and only when things are unknown invoke the `_complete` command).

dkarlovi commented:
@wouterj assuming 
1. Composer integrates this current work as-is
2. this means the `composer` app provides the `completion` command, ie. `composer completion bash` will work
3. the packager runs it, places the output script in the proper place

It should already work, no? The script will target the `composer` "binary", invoke the `_complete` command properly, etc. I don't see what special casing would actually be needed.

theofidry commented on src/Symfony/Component/Console/Completion/Completion.php:
```diff
@@ -0,0 +1,95 @@
+     */
+    public function suggestValues(array $values): self
+    {
+        $this->valueSuggestions += $values;
```
beware that this may override values due to common indexes. Is this desired?

wouterj replied:
Good catch, fixed

theofidry commented on src/Symfony/Component/Console/Completion/Completion.php:
```diff
@@ -0,0 +1,95 @@
+ *
+ * @author Wouter de Jong <wouter@wouterj.nl>
+ */
+class Completion
```
`CompletionSuggestion`? (might apply to the other classes)

theofidry commented on src/Symfony/Component/Console/Completion/CompletionInput.php:
```diff
@@ -0,0 +1,257 @@
+     */
+    public static function fromString(string $inputStr, int $currentIndex): self
+    {
+        preg_match_all('/(?<=^|\s)([\'"]?)(.+?)(?<!\\\\)\1(?=$|\s)/', $inputStr, $tokens);
```
might be worth adding a comment of what this regex is for/necessary?

wouterj replied:
I think that's already covered by the PHPdoc for this method. I've added some tests to further clarify the behavior of this regex

theofidry commented on src/Symfony/Component/Console/Completion/CompletionInput.php:
```diff
@@ -0,0 +1,257 @@
+
+            $option = $this->getOptionFromToken($optionToken);
+            if (null === $option && !$this->isCursorFree()) {
+                $this->completionType = self::TYPE_OPTION_NAME;
```
I tend to find this sort of API a bit brittle: it's easy to miss that all properties need to be set together. I am also not sure if this class is supposed to be extendable or not (it is not final at least, didn't check how hard/easy it would be to extend).

WDYT of adding a private accessor `setCompletionValues(string $type, string $anme, string $value)`?

wouterj replied:
I do agree with you that this a bit brittle, but this is indeed not meant to be extended (I've made the class final now - thanks to your comment). The same comment as https://github.com/symfony/symfony/pull/42251#discussion_r731222535 applies here imho.
Besides, we don't always set all 3 properties at the same time, sometimes only 1 or 2.

theofidry commented on src/Symfony/Component/Console/Completion/CompletionInput.php:
```diff
@@ -0,0 +1,257 @@
+    public const TYPE_OPTION_NAME = 'option_name';
+    public const TYPE_NONE = 'none';
+
+    private $tokens;
```
Not entirely sure what's the stance over this so asking in case: what about adding property types there?

wouterj replied:
We'll do that in 6.0 (see #43326)

theofidry commented on src/Symfony/Component/Console/Completion/CompletionInput.php:
```diff
@@ -0,0 +1,257 @@
+
+        $relevantToken = $this->getRelevantToken();
+        if ('-' === $relevantToken[0]) {
+            // the current token is an input option: complete either option name or option value
```
what about moving the different blocks into dedicated methods? e.g. `addCompletionForInputOption()`

wouterj replied:
As with all parsing, it's really horrible code (lots of nested if statements) but not so easy to clean-up.

In this case, `$relevantToken[0] === '-'` can have multiple meanings: Either it's just `-` or `--`, indicating the next tokens are all argument values. Or it's the start of an option (short or long). And things get even more messy with things like `-eprod`, `-vvv` and `-ev` (indicating respectively an option with its value, an option without value, and 2 options).

I also don't think it's worth cleaning this code up. All "messiness" is contained in a single method, which will not be touched regularly (how we parse options, arguments and values is pretty static).

theofidry reviewed (commented):
A few minor comments/nitpicks. A big 👍 for this feature

wouterj commented:
Thanks for the review @theofidry!

PR updated and rebased. We have a new failure (7.2) but I can't find any error logs whatsoever about this failure.

fabpot commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,117 @@
+
+        $this
+            ->setHelp(<<<EOH
+The <info>%command.name%</info> command dumps the shell completion script required
```
Small nitpick: we are using the short closing tags for internal commands:

```suggestion
The <info>%command.name%</> command dumps the shell completion script required
```

Can you fix all of them?

fabpot commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,117 @@
+
+Dump the script to a global completion file and restart your shell:
+
+    <info>%command.full_name% ${shell} | sudo tee /etc/bash_completion.d/${commandName}</info>
```
Does it work for zsh as well? Or do we need another example for it?

fabpot commented on src/Symfony/Component/Console/Command/CompleteCommand.php:
```diff
@@ -0,0 +1,195 @@
+    protected function configure(): void
+    {
+        $this
+            ->addOption('shell', 's', InputOption::VALUE_REQUIRED, 'The shell type (e.g. "bash" or "zsh")')
```
I think we need to remove zsh here as it's not supported AFAIU.

fabpot commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,117 @@
+<comment>Static installation
+-------------------</comment>
+
+Dump the script to a global completion file and restart your shell:
```
Maybe we should be explicit about the shells we support, which is only bash for now.

wouterj commented:
Thanks, PR updated.

fabpot commented:
Thank you @wouterj.

stof commented on src/Symfony/Component/Console/Command/DumpCompletionCommand.php:
```diff
@@ -0,0 +1,116 @@
+        if (!file_exists($debugFile)) {
+            touch($debugFile);
+        }
+        $process = new Process(['tail', '-f', $debugFile], null, null, null, 0);
```
what happens on windows (and yes, bash can be used on windows, which does not imply that `tail` is available in the PATH AFAICT) ?

dkarlovi replied:
This needs some more special casing. For example, you might not have `symfony/process` installed.

GromNaN commented on src/Symfony/Component/Console/Completion/CompletionInput.php:
```diff
@@ -0,0 +1,257 @@
+        return $parseOptions;
+    }
+
+    private function getOptionValueFromToken(string $optionToken): string
```
Is this function actually used?

GromNaN commented on src/Symfony/Component/Console/Completion/CompletionInput.php:
```diff
@@ -0,0 +1,257 @@
+
+    private $tokens;
+    private $currentIndex;
+    private $indexToArgumentIndex = [];
```
This property is not used.
